Emis

aub opened the pull request:
This adds support for the eMIS api. It's a SOAP API that we're using to access certified veteran status, but it also provides a bunch of other data.

Closes department-of-veterans-affairs/vets.gov-team#2040

robbiethegeek-usds commented:
@kreek and @saneshark please review and give feedback!

markolson commented on lib/emis/configuration.rb:
```diff
@@ -0,0 +1,44 @@
+
+module EMIS
+  class Configuration < Common::Client::Configuration::SOAP
+    SSL_CERT = begin
```
Between HCA, MVI and EMIS, we could pull this SSL_CERT/SSL_KEY load code into the SOAP class.

aub replied:
Good idea, I'll do that.

markolson commented on lib/emis/military_information_configuration.rb:
```diff
@@ -0,0 +1,18 @@
+
+module EMIS
+  class MilitaryInformationConfiguration < Configuration
+    URL = Settings.emis.military_information_url
```
Are these `URL` constants used?

aub replied:
I think they're somehow required by the base class, but I'll check it out to make sure

kreek commented on lib/emis/service.rb:
```diff
@@ -0,0 +1,68 @@
+        soapaction: "http://viers.va.gov/cdi/eMIS/#{operation.camelize(:lower)}/v1"
+      )
+      response_type.new(raw_response)
+      # :nocov:
```
This may need eMIS involvement but does their service return 'proper' http errors or can it also return XML defined service errors in 200 responses? (like MVI, EVSS do). Probably good idea to get some specs for invalid/error responses too.

saneshark commented on lib/emis/responses/response.rb:
```diff
@@ -0,0 +1,91 @@
+          end
+        end
+
+        OpenStruct.new(result_hash)
```
also not a big fan of using OpenStruct. I know Virtus is currently in search of a new maintainer and there are a bunch of alternatives to a coercion library like it, but I think it might be better than using open struct seeing as it is the pattern we've already used elsewhere.

In other words, we can use a parser middleware to get the data into a format where it can be deserialized / marshalled into a Virtus object. And Virtus has many coercion types that can then be used to coerce types. In addition, one can easily create his or her own coercion types, for example: Common::UTCTime

aub commented:
Ok, @kreek and @saneshark I removed the OpenStruct stuff in favor of using Virtus. It creates another set of classes to represent the models, which have another copy of the attribute names, but it seems like there are some benefits to doing it. It would be great to be able to write code that treats all of these responses in the same way, but the irregularity of the API makes this really difficult.

markolson commented:
I know I'm all "hey, pull this out into a common lib!" but the protected methods in response.rb look ripe for a common xml->virtus base class.

aub commented:
@markolson I see your point there. The thing is that it's really not much code and it necessarily contains some really annoying hacks that are specific to the way emis works. I can't decide if it's worth factoring it out and making it a little more complicated for the purpose of reusing that little bit of code.

markolson commented:
@aub since it's just single-use right now, it does make sense to keep it in just the Emis namespace until we know it could/should be re-used somewhere else. Though, I think those hacks are going to be pretty common to any SOAP endpoints, fwiw.

aub commented:
@kreek I also just added some specs for a couple of error conditions. It looks like they do return proper status codes when things go wrong.

With that, I think I've addressed all of the comments. @saneshark or @markolson, anything else or is this good to go?

kreek commented:
LGTM 🚀

aub commented:
Awesome, thanks @kreek
